test/realm_test.dart

```diff
@@ -1394,6 +1396,173 @@ Future<void> main([List<String>? args]) async {
+    expect(() => Realm.compact(config), throws<RealmException>("Can't compact an in-memory Realm"));
+  });
+
+  test('Realm - readonly realm can not be compacted', () async {
+    var path = p.join(Configuration.defaultStoragePath, "${generateRandomString(8)}.realm");
+    var config = Configuration.local([Product.schema], path: path);
+    final beforeCompactSize = await createRealmForCompact(config);
+
+    config = Configuration.local([Product.schema], isReadOnly: true, path: path);
+    expect(() => Realm.compact(config), throws<RealmException>("Can't compact a read-only Realm"));
+
+    //test the realm can be opened.
+    final realm = getRealm(config);
+  });
+
+  test('Realm - disconnected sync realm can be compacted', () async {
+    var config = Configuration.disconnectedSync([Product.schema], path: p.join(Configuration.defaultStoragePath, "${generateRandomString(8)}.realm"));
+
+    final beforeCompactSize = await createRealmForCompact(config);
+
+    final compacted = Realm.compact(config);
+
+    validateCompact(compacted, config.path, beforeCompactSize);
+
+    //test the realm can be opened.
+    final realm = getRealm(config);
+  });
+
+  baasTest('Realm - synced realm can be compacted', (appConfiguration) async {
+    final app = App(appConfiguration);
+    final credentials = Credentials.anonymous();
+    final user = await app.logIn(credentials);
+    final path = p.join(Configuration.defaultStoragePath, "${generateRandomString(8)}.realm");
+    final config = Configuration.flexibleSync(user, [Product.schema], path: path);
+    final beforeCompactSize = await createRealmForCompact(config);
+    user.logOut();
+    Future<void>.delayed(Duration(seconds: 5));
+
+    final compacted = Realm.compact(config);
+
+    validateCompact(compacted, config.path, beforeCompactSize);
+
+    //test the realm can be opened.
+    final realm = getRealm(Configuration.disconnectedSync([Product.schema], path: path));
+  });
+
+  baasTest('Realm - synced encrypted realm can be compacted', (appConfiguration) async {
+    final app = App(appConfiguration);
+    final credentials = Credentials.anonymous();
+    final path = p.join(Configuration.defaultStoragePath, "${generateRandomString(8)}.realm");
+    final user = await app.logIn(credentials);
+    List<int> key = List<int>.generate(encryptionKeySize, (i) => random.nextInt(256));
+    final config =
+        Configuration.flexibleSync(user, [Product.schema], encryptionKey: key, path: path);
+    final beforeCompactSize = await createRealmForCompact(config);
+    user.logOut();
+    Future<void>.delayed(Duration(seconds: 5));
+
+    final compacted = Realm.compact(config);
+
+    validateCompact(compacted, config.path, beforeCompactSize);
+
+    //test the realm can be opened.
+    final realm = getRealm(Configuration.disconnectedSync([Product.schema], path: path, encryptionKey: key));
```

Comment:
I would add `expect(realm.isClosed, false);` at end of each test, but it is not so important.